### ISSUE TITLE
how to write the predict python code in  BucketingModule with variable-length sequence?

### ISSUE THREAD

**Issue by sunspring320** [dmlc/mxnet#6117]
in my project, i try my project with cnn+lstm+ctc net
i start my module with mod 
 mx.mod.BucketingModule(sym_gen, default_bucket_key=data_train.default_bucket_key, context=contexts)
 and the model is train succeed, but in prediction stage, i try 
mx.model.load_checkpoint(modelPrefix, 50) :Cannot find Operator WarpCTC in registry
mxnet_predict   Predictor  : "'Predictor' object has no attribute 'handle'" in <bound method Predictor.__del__ of <mxnet_predict.Predictor object at 0x7f749cf97a10>> ignored
but all not succeed,
how do i wirte the bucketing prediction python code ?
thanks!

**Comment by Godricly** [comment]
When you do prediction, you can replace it with softmax and do some post processing. The predicting is almost the same as the training.

**Comment by sunspring320** [comment]
@Godricly   i write my prediction follow your code as the same as Bucketing module prediction #5008 
def load_param(self, param_path):
        node_data = mx.nd.load(param_path)
        # pickle is not working although the doc says so
        # param_file = open('./model/card-0001.params','rb')
        # print pickle.load(param_file)

        # parse node data into 2 lists
        arg_param={}
        aux_param={}
        for k, v in node_data.items():
            tp, name = k.split(':', 1)
            if tp == 'arg':
                arg_param[name] = v
            if tp == 'aux':
                aux_param[name] = v
        print(arg_param)
        print(aux_param)
        return arg_param, aux_param

    def __init_ocr(self):
        batch_size = 1
        self.predictor=mx.module.BucketingModule(sym_gen, default_bucket_key=self.default_bucket_key,context=mx.context.gpu(0))
        arg_param, aux_param = self.load_param(self.path_of_params)
        self.predictor.bind(self.data_shapes, self.label_shapes,for_training=False)
        self.predictor.init_params(arg_params=arg_param, aux_params=aux_param)

    def predict(self, data_batch):
        self.predictor.forward(data_batch, is_train=False)
        return self.predictor.get_outputs()

    # sym_gen
    def sym_gen(bucket_key):
        label_length, lstm_length = bucket_key.split(',')
        lstm_length = int(lstm_length)
        label_length = int(label_length)
        return get_symbol(num_classes, num_hidden, lstm_length, label_length), data_name, label_name
but i can not load my trained model, would you provide me the full predict code or some help?
what the define of data_shapes, label_shapes,bucket_key，default_bucket_key?

**Comment by Godricly** [comment]
What is the error you have now and can you post your symbol here?

For those parameters, it depends on how you compute them. What is you input image size? 
When you do predict you actually don't know the label shape. The default bucket key is used in training to allocate the memory. And bucket_key describes the shape your data. 

For example. if you want to predict a image of shape 300x30 with 100x1 as the input shape for lstm. You can have data_shape as 300x30 and bucket_key of 100 which tells how long the lstm is. The CTC is replaced with softmax in prediction.

**Comment by sunspring320** [comment]
@Godricly thans, my symbol as follow:
#-*- coding: UTF-8 -*-
# pylint:skip-file
import sys
sys.path.insert(0, "../../python")
import mxnet as mx
import numpy as np
from collections import namedtuple
import time
import math
LSTMState = namedtuple("LSTMState", ["c", "h"])
LSTMParam = namedtuple("LSTMParam", ["i2h_weight", "i2h_bias",
                                     "h2h_weight", "h2h_bias"])
LSTMModel = namedtuple("LSTMModel", ["rnn_exec", "symbol",
                                     "init_states", "last_states",
                                     "seq_data", "seq_labels", "seq_outputs",
                                     "param_blocks"])

def lstm(num_hidden, indata, prev_state, param, seqidx, layeridx, dropout=0.):
    """LSTM Cell symbol"""
    if dropout > 0.:
        indata = mx.sym.Dropout(data=indata, p=dropout)
    i2h = mx.sym.FullyConnected(data=indata,
                                weight=param.i2h_weight,
                                bias=param.i2h_bias,
                                num_hidden=num_hidden * 4,
                                name="t%d_l%d_i2h" % (seqidx, layeridx))
  
    h2h = mx.sym.FullyConnected(data=prev_state.h,
                                weight=param.h2h_weight,
                                bias=param.h2h_bias,
                                num_hidden=num_hidden * 4,
                                name="t%d_l%d_h2h" % (seqidx, layeridx))
    gates = i2h + h2h
    slice_gates = mx.sym.SliceChannel(gates, num_outputs=4,
                                      name="t%d_l%d_slice" % (seqidx, layeridx))
    in_gate = mx.sym.Activation(slice_gates[0], act_type="sigmoid")
    in_transform = mx.sym.Activation(slice_gates[1], act_type="tanh")
    forget_gate = mx.sym.Activation(slice_gates[2], act_type="sigmoid")
    out_gate = mx.sym.Activation(slice_gates[3], act_type="sigmoid")
    next_c = (forget_gate * prev_state.c) + (in_gate * in_transform)
    next_h = out_gate * mx.sym.Activation(next_c, act_type="tanh")
    return LSTMState(c=next_c, h=next_h)
#for train
def lstm_unroll(num_lstm_layer, seq_len, num_hidden, num_class , num_label , dropout=0.):
    #print("seq_len: %d", seq_len)
    #embed_weight = mx.sym.Variable("embed_weight")
    cls_weight = mx.sym.Variable("cls_weight")
    cls_bias = mx.sym.Variable("cls_bias")
    param_cells = []
    last_states = []
    for i in range(num_lstm_layer):
        param_cells.append(LSTMParam(i2h_weight=mx.sym.Variable("l%d_i2h_weight" % i),
                                     i2h_bias=mx.sym.Variable("l%d_i2h_bias" % i),
                                     h2h_weight=mx.sym.Variable("l%d_h2h_weight" % i),
                                     h2h_bias=mx.sym.Variable("l%d_h2h_bias" % i)))
        state = LSTMState(c=mx.sym.Variable("l%d_init_c" % i),
                          h=mx.sym.Variable("l%d_init_h" % i))
        last_states.append(state)
    assert(len(last_states) == num_lstm_layer)
    data = mx.sym.Variable('data')
    label = mx.sym.Variable('softmax_label')
    #arg_shape, output_shape, aux_shape = data.infer_shape(data=(32,65,128))
    #arg_shape, output_shape, aux_shape = label.infer_shape(data=(32, 9))
    #print(output_shape)#[(32L, 512L, 1L, 25L)]
    wordvec = mx.sym.SliceChannel(data=data, num_outputs=seq_len, squeeze_axis=1)
    hidden_all = []
    print 'seq_len:', seq_len
    for seqidx in range(seq_len):
        hidden = wordvec[seqidx]
        # stack LSTM，
        for i in range(num_lstm_layer):
            if i == 0:
                dp_ratio = 0.
            else:
                dp_ratio = dropout
            next_state = lstm(num_hidden, indata=hidden,
                              prev_state=last_states[i],
                              param=param_cells[i],
                              seqidx=seqidx, layeridx=i, dropout=dp_ratio)
            hidden = next_state.h
            last_states[i] = next_state
        # decoder
        if dropout > 0.:
            hidden = mx.sym.Dropout(data=hidden, p=dropout)
        hidden_all.append(hidden)

    hidden_concat = mx.sym.Concat(*hidden_all, dim=0)
    pred = mx.sym.FullyConnected(data=hidden_concat, num_hidden=num_class,
                                 weight=cls_weight, bias=cls_bias, name='pred')

    #label = mx.sym.transpose(data=label)
    #label = mx.sym.Reshape(data=label, target_shape=(0,))
    label = mx.sym.Reshape(data=label, shape=(-1,))
    label = mx.sym.Cast(data = label, dtype = 'int32')
    #print'ctc'
    sm = mx.sym.WarpCTC(data=pred, label=label, label_length = num_label, input_length = seq_len)
    #print'ctc___end'
    return sm
#predict symbol 
def lstm_inference_symbol(num_lstm_layer, lstm_length,
                          num_hidden ,label_length, num_classes, dropout=0.):
    seqidx = 0
    cls_weight = mx.sym.Variable("cls_weight")
    cls_bias = mx.sym.Variable("cls_bias")
    param_cells = []
    last_states = []
    for i in range(num_lstm_layer):
        param_cells.append(LSTMParam(i2h_weight = mx.sym.Variable("l%d_i2h_weight" % i),
                                      i2h_bias = mx.sym.Variable("l%d_i2h_bias" % i),
                                      h2h_weight = mx.sym.Variable("l%d_h2h_weight" % i),
                                      h2h_bias = mx.sym.Variable("l%d_h2h_bias" % i)))
        state = LSTMState(c=mx.sym.Variable("l%d_init_c" % i),
                          h=mx.sym.Variable("l%d_init_h" % i))
        last_states.append(state)
    assert(len(last_states) == num_lstm_layer)
    data = mx.sym.Variable("data")
    # stack LSTM
    for i in range(num_lstm_layer):
        if i==0:
            dp=0.
        else:
            dp = dropout
        next_state = lstm(num_hidden, indata=data,
                          prev_state=last_states[i],
                          param=param_cells[i],
                          seqidx=seqidx, layeridx=i, dropout=dp)
        data = next_state.h
        last_states[i] = next_state
    # decoder
    if dropout > 0.:
        data = mx.sym.Dropout(data=data, p=dropout)

    fc = mx.sym.FullyConnected(data=data, num_hidden=num_classes,
                               weight=cls_weight, bias=cls_bias, name='pred')
    sm = mx.sym.SoftmaxOutput(data=fc, name='softmax')
    #sm = mx.sym.WarpCTC(data=fc, label=label, label_length=num_label, input_length=input_size)
    output = [sm]
    for state in last_states:
        output.append(state.c)
        output.append(state.h)
    return mx.sym.Group(output)

this is my symbol , how can i  predict my data?
my data_shape is (bucket_lenght，128), label_shapes is (1,9), default_bucket_key is 65.
thanks!

**Comment by Godricly** [comment]
1. What is your error when trying to load the model with my code?  Can you show me that log?
2. Are you able to load the parameters from model files?
3. Another thing I notice is that you are using old LSTM interface. Maybe you can try the new one which is more convinient.
4. As I mentioned more than once, the warpCTC is replaced with Softmax when you do testing. The good thing is that it has no parameters.

**Comment by Godricly** [comment]
The bucket key of my code was contains two parts. 1st is the warpCTC label length, 2nd is the lstm sequence length. In testing, since you don't know the label length. You bucket_key only represents the lstm sequence length. I suggest you to write another symbol for testing but shares same symbol names for those layers with parameters.

**Comment by sunspring320** [comment]
@Godricly i load the parameter successful, but in the step of 
self.predictor.bind(self.data_shapes, self.label_shapes,for_training=False)
in BucketingModule  data_shapes is a vary length  sequence, what data_shapes for the input of bind?

**Comment by sunspring320** [comment]
@Godricly thans, i think i finished the problem! thanks very much !

**Comment by yhl41001** [comment]
@sunspring320 i have the same problem, in project https://github.com/sunspring320/mxnet-cnn-lstm-ctc-ocr, how to construct the predict code, have you finished?
